metsim/metsim.py

```diff
@@ -257,80 +253,15 @@ def launch(self):
-
-            # Add in some end point data for continuity in chunking
-            times_ext = times.union([times[0] - pd.Timedelta("1 days"),
-                                     times[-1] + pd.Timedelta("1 days")]
-                                    ).intersection(time_dim)
-            data = self.met_data.sel(time=times_ext)
-            self.setup_output(self.met_data.sel(time=times))
-            for site in self.site_generator:
-                locs = {k: v for k, v in zip(self.params['iter_dims'], site)}
-                # Don't run masked cells
-                if not self.domain['mask'].sel(**locs).values > 0:
-                    continue
-                stat = self.pool.apply_async(
-                    wrap_run,
-                    args=(self.method.run, locs, self.params,
-                          data.sel(**locs),
-                          self.state.sel(**locs),
-                          self.disagg, times, label),
-                    callback=self._unpack_results)
-                status.append(stat)
-            self.pool.close()
-            # Check that everything worked
-            [stat.get() for stat in status]
-            self.pool.join()
-            self.write(label)
-
-    def run(self):
-        """
-        Kicks off the disaggregation and queues up data for IO
-        """
-        self._validate_setup()
-        iter_list = [self.met_data[dim].values
-                     for dim in self.params['iter_dims']]
         self.disagg = int(self.params['time_step']) < cnst.MIN_PER_DAY
         self.method = MetSim.methods[self.params['method']]
 
-        time_dim, groups = self._get_time_dim_and_time_dim()
+        out = self.method.run(self.met_data)
 
-        for label, times in groups.items():
-            logger.info("Beginning {}".format(label))
-            self.site_generator = itertools.product(*iter_list)
-            # Add in some end point data for continuity
-            times_ext = times.union([times[0] - pd.Timedelta("1 days"),
-                                     times[-1] + pd.Timedelta("1 days")]
-                                    ).intersection(time_dim)
-            data = self.met_data.sel(time=times_ext)
-            self.setup_output(self.met_data.sel(time=times))
-
-            for site in self.site_generator:
-                locs = {k: v for k, v in zip(self.params['iter_dims'], site)}
-                # Don't run masked cells
-                if not self.domain['mask'].sel(**locs).values > 0:
-                    continue
-                wrap_results = wrap_run(
-                    self.method.run, locs, self.params, data.sel(**locs),
-                    self.state.sel(**locs), self.disagg, times, label)
-
-                # Cut the returned data down to the correct time index
-                self._unpack_results(wrap_results)
-
-            self.write(label)
+        if self.disagg:
+            self.output = disaggregate(out)
+        else:
+            self.output = out
```

Comment:
More logic to come here. Specifically, we don't have any notion of chunked output (per year files). We also don't have any logic yet for specifying dask chunking schemes.

Reply:
@jhamman just FYI, Kristen Whitney here at ASU has created a python script to loop over years and call MetSim on individual monthly forcing files and concatenate the output into yearly files...  It's PR #106 ...  Will the dask feature supercede this?

Reply:
@tbohn - I think we can handle that in #125.